### PR TITLE
Add dynamodb stream

### PR DESCRIPTION
**Pull request by humulla**
Add Dynamodb stream to the dynamodb table, this will keep a record of any attributes that have been updated.
This is for the MMH-HFS sync, so whenever the tenure end date is updated the dynamodb stream will record it and then trigger the finance-dynamodb-stream lambda function that would then update the relevant finance tables.

### CODE REVIEW

File: terraform/development/main.tf [hunk 1]
```diff
@@ -142,5 +142,12 @@ data "aws_iam_policy_document" "sns_topic_policy" {
...
     }	
-}	
+}
+
+# TO DO: Add lambda function name once created
+# resource "aws_lambda_event_source_mapping" "aws_lambda_event_source" {
+#   event_source_arn  = aws_dynamodb_table.TenureInformation.stream_arn
+#   function_name     = aws_lambda_function.lambda_function.arn
+#   starting_position = "LATEST"
+# }
```
Comment (Duslerke) [inline terraform/development/main.tf:152]: Could add a simple lambda deployed as part of the TenureInformation API.

(add a simple lambda that logs "hello world" to Tenure Information APIs serverless.yaml and then use its ARN here)

Would be nice to have the lambda as part of this PR sort of as a proof of concept that we could check in the cloudwatch logs. :thinking: 

Idk, what do you think?

**Review by LBHCallumM — DISMISSED** [review]

**Review by Duslerke — COMMENTED** [review]
It appears the that the generic `lambdaExecutionRole` isn't enough for the use of this lambda by the dynamo stream event.

You'd need to give that `dynamodb-stream-poc` lambda the ability to do the GetRecords, GetShardIterator, DescribeStream, and ListStreams actions. This either needs to be defined as a new Role within the serverless file, or be edited manually via AWS UI console _(since this is merely a prototype)_.

**Review by Duslerke — COMMENTED** [review]
Rename dynamo permission actions:
```
dynamodb:DescribeStream
dynamodb:GetRecords
dynamodb:GetShardIterator
dynamodb:ListStreams
```

File: TenureInformationApi/serverless.yml [hunk 2]
```diff
@@ -112,6 +116,22 @@ resources:
...
+                    - "iam:DescribeStream"
+                    - "iam:ListStreams"
+                    - "iam:GetRecords"  
+                    - "iam:GetShardIterator"
+                    - "dynamodb:DescribeStream"
+                    - "dynamodb:ListStreams"
+                    - "dynamodb:GetRecords"
+                    - "dynamodb:GetShardIterator"
+
+                  Resource: "*"
```
Comment (Duslerke) [inline TenureInformationApi/serverless.yml:134]: This wildcard is temporary for the purposes of proof of concept.
It was agreed that once the real lambda gets created, the dynamo permissions will be set against the stream: `arn:aws:dynamodb:eu-west-2:364864573329:table/TenureInformation/stream/*`

**Review by Duslerke — DISMISSED** [review]

**Review by Duslerke — DISMISSED** [review]

File: terraform/production/dynamodb.tf [hunk 3]
```diff
@@ -17,3 +19,27 @@ resource "aws_dynamodb_table" "tenureinformationapi_dynamodb_table" {
     enabled = true
   }
 }
+
+data "aws_iam_policy_document" "assume_role" {
```
Comment (FaisalHackney) [inline terraform/production/dynamodb.tf:23]: Does the apply of this terraform config on the Prod DynamoDB happen straightaway or is there an manual approval gate it hits in the CI/CD pipeline?

Comment (humulla) [reply terraform/production/dynamodb.tf:23]: there is a manual approval and it only happens on the release branch

**Review by FaisalHackney — APPROVED** [review]

**Pull request merged**